File: 6-class-hierarchy/Object.c
#include <stdio.h>
#include <stdarg.h>
#include <stddef.h>
#include <malloc.h>
#include <assert.h>

#include "Object.r"
#include "Object.h"


/**
 * "public" definitions
 */

const void * classOf (const void * _self) {
    const struct Object * self = _self;

    assert(self && self -> class);
    return self -> class;
}

size_t sizeOf (const void * _self) {
    const struct Class * class = classOf(_self);

    return class -> size;
}


/**
 * Object-specific definitions
 */

static void * Object_ctor (void * _self, va_list * app) {
    return _self;
}

static void * Object_dtor (void * _self) {
    return _self;
}

static int Object_differ (const void * _self, const void * _other) {
    return _self == _other;        /* like that ? */
}

static int Object_puto (const void * _self, FILE * fp) {
    const struct Class * class = classOf(_self);

    return fprintf(fp, "%s at %p\n", class -> name, _self);
}

File: 6-class-hierarchy/Object.r
#ifndef OBJECT_R
#define OBJECT_R

#include <stdio.h>
#include <stddef.h>
#include <stdarg.h>


struct Object {
    const struct Class * class;    /* object's description */
};

struct Class {
    const struct Object _;         /* class' description */
    const char * name;             /* class' name */
    const struct Class * super;    /* class' super class */
    size_t size;                   /* class' object's size */
    void * (* ctor)   (void * self, va_list * app);
    void * (* dtor)   (void * self);
    int    (* differ) (const void * self, const void * other);
    int    (* puto)   (const void * self, FILE * fp);
};


#endif        /* OBJECT_R */

